server/client.go

```diff
@@ -2099,7 +2113,8 @@ func (c *client) processConnect(arg []byte) error {
 		}
 
 		// If no account designation.
-		if c.acc == nil {
+		// Do this only for CLIENT and LEAF connections.
+		if c.acc == nil && (c.kind == CLIENT || c.kind == LEAF) {
```

Comment:
This was the main issue with the Gateway reported memory leak. The connection was assigned to the global account's client list, but never removed from there. So I believe that we should do it only for those type of connections.